### PR TITLE
Refactor aggregates

### PR DESCRIPTION
**Pull request by devyetii**
## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  |   |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  |   |
| zwalletcli  |   |
| 0chain  | fix/refector-aggregates  |
| blobber  |   |
| authorizer  |   |
| 0box  |   |
| 0dns  |   |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  |   |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  |   |
| zwalletcli  |   |
| 0chain  | sprint-july-3  |
| blobber  |   |
| authorizer  |   |
| 0box  |   |
| 0dns  |   |

File: code/go/0chain.net/smartcontract/dbs/event/provider.go [hunk 1]
```diff
@@ -108,6 +122,311 @@ func (edb *EventDb) ReplicateProviderAggregates(round int64, limit int, offset i
...
+
+func getProvidersById[P IProvider](edb *EventDb, ids []string) ([]IProvider, error) {
+	var (
+		model P
+		providers []P
+		tableName = model.TableName()
+	)
+	err := edb.Get().
+		Model(&model).
+		Where(fmt.Sprintf("%v.id IN (?)", tableName), ids).
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/provider.go]: do not use `IN` use temp table or `unnest` directly

File: code/go/0chain.net/smartcontract/dbs/event/provider.go [hunk 2]
```diff
@@ -108,6 +122,311 @@ func (edb *EventDb) ReplicateProviderAggregates(round int64, limit int, offset i
...
+	return nil, common.NewError("invalid_provider_type", "invalid provider type")
+}
+
+func getProvidersById[P IProvider](edb *EventDb, ids []string) ([]IProvider, error) {
+	var (
+		model P
+		providers []P
+		tableName = model.TableName()
+	)
+	err := edb.Get().
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/provider.go:195]: how many providers can we have here? should we batch the results to save the memory?

File: code/go/0chain.net/smartcontract/dbs/event/provider.go [hunk 3]
```diff
@@ -108,6 +122,311 @@ func (edb *EventDb) ReplicateProviderAggregates(round int64, limit int, offset i
...
+
+	idsLists := map[spenum.Provider][]string{
+		spenum.Blobber:    make([]string, 0, len(ids[spenum.Blobber])),
+		spenum.Miner:      make([]string, 0, len(ids[spenum.Miner])),
+		spenum.Sharder:    make([]string, 0, len(ids[spenum.Sharder])),
+		spenum.Authorizer: make([]string, 0, len(ids[spenum.Authorizer])),
+		spenum.Validator:  make([]string, 0, len(ids[spenum.Validator])),
+	}
+	
+	for provider, ids := range ids {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/provider.go]: maybe rename ids and ids

**Review by dabasov — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/dbs/event/provider.go (hunk 2 above)
Comment (devyetii) [reply code/go/0chain.net/smartcontract/dbs/event/provider.go:195]: Not much, depends on the number of events in one round.

**Review by dabasov — APPROVED** [review]
lgtm

**Pull request merged**